archive/imo/imo1975_q1.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2022 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+import data.real.basic
+import data.nat.interval
+import algebra.order.rearrangement
+import algebra.big_operators.ring
+
+/-!
+# IMO 1975 Q1
+
+Let `x₁, x₂, ... , xₙ` and `y₁, y₂, ... , yₙ` be two sequences of real numbers, such that
+`x₁ ≥ x₂ ≥ ... ≥ xₙ` and `y₁ ≥ y₂ ≥ ... ≥ yₙ`. Prove that if `z₁, z₂, ... , zₙ` is any permutation
+of `y₁, y₂, ... , yₙ`, then
+`∑ i in finset.Icc 1 n, (x i - y i) ^ 2 ≤ ∑ i in finset.Icc 1 n, (x i - z i) ^ 2`
+
+# Solution
+
+Firstly, we expand the squares withing both sums and distribute into separate finite sums. Then,
+noting that `∑ i in finset.Icc 1 n, (y i) ^ 2 = ∑ i in finset.Icc 1 n, (z i) ^ 2`, it remains to
+prove that `∑ i in finset.Icc 1 n, x i * z i ≤ ∑ i in finset.Icc 1 n, x i * y i`, which is true by
+the Rearrangement Inequality
+-/
+
+open_locale big_operators
+
+/-- Let `n` be a natural number, `x` and `y` be as in the problem statement and `σ` be the
+permutation of natural numbers such that `z = y ∘ σ` -/
+variables (n : ℕ) (x y : ℕ → ℝ) (σ : equiv.perm ℕ) (hx : antitone_on x (finset.Icc 1 n))
```

Comment:
I'm not very familiar with the permutation API, but is it preferable to make `σ` a permutation of `ℕ` instead of one of `fin n`?

Reply:
I think in an ideal world, I would prefer a permutation of `fin n`, yet if we are stating the goal using `finset.sum`, then we have our variables in `nat` and the coercion to `fin n` I have found to be quite painful.

Reply:
It might be nice to have both statements, since we likely want to have the API to make doing this with `fin` pleasant too. But the current proof looks very nice.